monitoring/collector/etcd.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/satellite/monitoring"
+	"github.com/gravitational/trace"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ETCDCollector collects ETCD stats from the given server and exports them using
+// the prometheus metrics package.
+type ETCDCollector struct {
+	client *roundtrip.Client
+	config monitoring.ETCDConfig
+	mutex  sync.RWMutex
+
+	isRunning typedDesc
+	health    typedDesc
+	isLeader  typedDesc
+}
+
+// NewETCDCollector returns an initialized ETCDCollector.
+func NewETCDCollector(config *monitoring.ETCDConfig) (Collector, error) {
+	transport, err := config.NewHTTPTransport()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(config.Endpoints) == 0 {
+		return nil, trace.BadParameter("no ETCD endpoints configured")
+	}
+
+	client, err := newClient(config.Endpoints[0], roundtrip.HTTPClient(&http.Client{
+		Transport: transport,
+		Timeout:   collectMetricsTimeout,
+	}))
+
+	return &ETCDCollector{
+		client: client,
+		config: *config,
+		isRunning: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "etcd", "up"),
+			"Whether scraping ETCD metrics was successful.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		health: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "etcd", "health"),
+			"Health status of ETCD.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		isLeader: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "etcd", "leader"),
+			"Endpoint is leader of ETCD cluster.",
+			nil, nil,
+		), prometheus.GaugeValue},
+	}, nil
+}
+
+// Collect implements prometheus.Collector.
+func (e *ETCDCollector) Collect(ch chan<- prometheus.Metric) error {
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	var m prometheus.Metric
+	var err error
+	success := true
+
+	resp, err := e.client.Get(e.client.Endpoint("v2", "stats", "leader"), url.Values{})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if bytes.Contains(resp.Bytes(), []byte("not current leader")) {
```

Comment:
this whole logic block should be refactored to:

```
// this is your default value, returned by default if all other nested blocks failed
metric, err = e.health.newConstMetric(0.0)
if err != nil {
  // nothing to do, return
  return
}

// then refactor logic to emit 1 if etcd is healthy, otherwise return the default metric
healthy, _ := e.healthStatus()
if !bytes.Contains(resp.Bytes(), []byte("not current leader")) && healthy {
   metric, err = e.health.newConstMetric(1.0)
}

ch <- metric

```

Reply:
I can't use default metric here, because it is a constant metric. Prometheus documentation says about using constant metrics for situation when you should scrape them from outside. I can switch to non-constant metric and then code with defaults will work(what I did before, but then switched to constant metrics).

Reply:
I feel there's too much noise around using constant metric values - including the metrics constructor giving an error back for a precondition test which needs to be checked on every metric generation.
From what I see, there're just two types of constant metrics used: `off` and `on` (subject to interpretation).
I'd rather have them declared as package locals:

[collector/etcd.go]
```go
// ETCDCollector collects ETCD stats from the given server and exports them using
// the prometheus metrics package.
type ETCDCollector struct {
	client *roundtrip.Client
	config monitoring.ETCDConfig
}

// NewETCDCollector returns an initialized ETCDCollector.
func NewETCDCollector(config *monitoring.ETCDConfig) (Collector, error) {
	transport, err := config.NewHTTPTransport()
	if err != nil {
		return nil, trace.Wrap(err)
	}

	if len(config.Endpoints) == 0 {
		return nil, trace.BadParameter("no ETCD endpoints configured")
	}

	client, err := newClient(config.Endpoints[0], roundtrip.HTTPClient(&http.Client{
		Transport: transport,
		Timeout:   collectMetricsTimeout,
	}))

	return &ETCDCollector{
		client: client,
		config: *config,
	}, nil
}

// Collect implements prometheus.Collector.
func (e *ETCDCollector) Collect(ch chan<- prometheus.Metric) error {
	resp, err := e.client.Get(e.client.Endpoint("v2", "stats", "leader"), url.Values{})
	if err != nil {
		return trace.Wrap(err)
	}

	if !isLeaderResp(resp.Bytes()) {
		ch <- leaderOff
		ch <- runningOn
		return nil
	}

	healthy, err := e.healthStatus()
	if err != nil {
		return trace.Wrap(err)
	}

	ch <- leaderOn
	if healthy {
		ch <- healthyOn
	} else {
		ch <- healthyOff
	}

	ch <- runningOn

	return nil
}

func isLeaderResp(resp []byte) bool {
	return bytes.Contains(resp, []byte("not current leader"))
}

var isRunning = typedDesc{prometheus.NewDesc(
	prometheus.BuildFQName(namespace, "etcd", "up"),
	"Whether scraping ETCD metrics was successful.",
	nil, nil,
), prometheus.GaugeValue}

var runningOn = isRunning.mustNewConstMetric(1.0)

var healthy = typedDesc{prometheus.NewDesc(
	prometheus.BuildFQName(namespace, "etcd", "health"),
	"Health status of ETCD.",
	nil, nil,
), prometheus.GaugeValue}

var healthyOn = healthy.mustNewConstMetric(1.0)
var healthyOff = healthy.mustNewConstMetric(0.0)

var isLeader = typedDesc{prometheus.NewDesc(
	prometheus.BuildFQName(namespace, "etcd", "leader"),
	"Endpoint is leader of ETCD cluster.",
	nil, nil,
), prometheus.GaugeValue}

var leaderOn = isLeader.mustNewConstMetric(1.0)
var leaderOff = isLeader.mustNewConstMetric(0.0)
```